### PR TITLE
Update conda packages to clang 11

### PR DESCRIPTION
**Pull request by cticenhour**
Updates moose-mpich conda recipe to use clang_osx-64 11.0.0 build hb91bd55_5 for new SDK support and updates related dependencies as a result of that change. 

Still to do:
- [x] Percolate changes to moose-mpich to other packages
- [x] Fixup new warnings in libMesh (see libmesh/libmesh#2812), gtest, and MOOSE
- [x] ~~Update any needed Linux dependencies~~ EDIT: none needed

Tagging @milljm 

Refs #16291

### CODE REVIEW

**Comment by cticenhour** [comment]
@lindsayad A libmesh update will need to be performed for this PR after libmesh/libmesh#2812 is finished and merged. Would you rather do that in a separate PR or as part of this?

File: framework/contrib/gtest/gtest/gtest.h
Comment (lindsayad) [inline framework/contrib/gtest/gtest/gtest.h:62]: Will this lead to ignoring bad copies in MOOSE unit test code? I'm guessing so

Comment (cticenhour) [reply framework/contrib/gtest/gtest/gtest.h:62]: I didn't consider that - this has been a longstanding issue in gtest, so I am unsure how to handle it short of diving into that code and fixing up our own copy.

File: framework/src/postprocessors/NumRelationshipManagers.C
Comment (lindsayad) [inline framework/src/postprocessors/NumRelationshipManagers.C:49]: `rm` is a pointer, so let's instead do `&` -> `* const`

File: framework/src/postprocessors/NumRelationshipManagers.C
Comment (lindsayad) [inline framework/src/postprocessors/NumRelationshipManagers.C:55]: `&` -> `* const`

File: framework/src/postprocessors/NumRelationshipManagers.C
Comment (lindsayad) [inline framework/src/postprocessors/NumRelationshipManagers.C:61]: `&` -> `* const`

**Review by lindsayad — DISMISSED** [review]
I love these compiler updates. They elucidate bad code

File: framework/src/postprocessors/NumRelationshipManagers.C
Comment (cticenhour) [reply framework/src/postprocessors/NumRelationshipManagers.C:49]: Oops - I got lazy and followed the compiler's suggestion. 😅 Thanks!

File: framework/contrib/gtest/gtest/gtest.h
Comment (lindsayad) [reply framework/contrib/gtest/gtest/gtest.h:62]: It depends on when the warnings happen. There are so many macros with gtest that *a priori*, without actually incurring the warnings during a build, it's impossible to know what macros spawn warnings. For instance, if this
```c++

TEST(ADFparserTest, JITCompile)
{
```
spawns a warning it would be obnoxious to put this diagnostic ignored pragma around every one of the macro instances.

Comment (lindsayad) [reply framework/contrib/gtest/gtest/gtest.h:62]: I don't know. What you have is probably ok. We should just remove this as soon as gtest fixes their shizz

**Comment by lindsayad** [comment]
> @lindsayad A libmesh update will need to be performed for this PR after [libMesh/libmesh#2812](https://github.com/libMesh/libmesh/pull/2812) is finished and merged. Would you rather do that in a separate PR or as part of this?

Probably separate

File: framework/contrib/gtest/gtest/gtest.h
Comment (cticenhour) [reply framework/contrib/gtest/gtest/gtest.h:62]: I could always temporarily remove and see what happens when I actually run the test suite. Do you know if we have `-Werror` turned on for that?

File: framework/src/postprocessors/NumRelationshipManagers.C
Comment (cticenhour) [reply framework/src/postprocessors/NumRelationshipManagers.C:49]: Actually, clang didn't like that extra `const`, since the type ends up as `const std::shared_ptr<RelationshipManager>`, so I'll stick with just a `*`.

Comment (cticenhour) [reply framework/src/postprocessors/NumRelationshipManagers.C:49]: Clang isn't happy with either `*` or `* const`:

```
/Users/icenct/projects/moose/framework/src/postprocessors/NumRelationshipManagers.C:49:23: error: variable 'rm' with type 'const auto *' has incompatible initializer of type 'const std::__1::shared_ptr<RelationshipManager>'
    for (const auto * rm : rms)
                      ^  ~
```

Shouldn't this declared variable be a reference to that shared pointer?

File: framework/contrib/gtest/gtest/gtest.h
Comment (cticenhour) [reply framework/contrib/gtest/gtest/gtest.h:62]: I'm actually a little tempted to fix this up myself - we're already a decent amount behind HEAD and it looks like gtest doesn't accept PRs. I'll ponder over the holiday :D

Comment (cticenhour) [reply framework/contrib/gtest/gtest/gtest.h:62]: Well perhaps not.....I thought we had forked gtest, but it seems as though the last change to this file was a superficial one.

File: framework/src/postprocessors/NumRelationshipManagers.C
Comment (lindsayad) [reply framework/src/postprocessors/NumRelationshipManagers.C:49]: Then I've led you astray. Since it's a `shared_ptr`, then the `&`s are good

**Review by lindsayad — COMMENTED** [review]
`shared_ptr ` instead of raw pointer...my concerns are invalid :smile:

**Comment by lindsayad** [comment]
Omg I have to explicitly dismiss my requested changes if I don't explicitly approve :-/

**Comment by cticenhour** [comment]
FYI - the changes in libMesh have been merged into master.

**Comment by cticenhour** [comment]
@milljm Did you want to do any updating with the Linux dependencies here? Or just leave gcc etc as-is since this is a clang-centric change due to Xcode 12?

**Comment by milljm** [comment]
The recipe changes look fine. But I would hold off until we fix firewall rules on civet (anaconda channel will never get updated with these new packages).

**Comment by milljm** [comment]
I am seeing a lot of warning while using LLVM 11 (in Spack). Not sure how things look on Conda.

**Comment by cticenhour** [comment]
I added some warnings to libMesh `ignore_warnings.h` that were popping up in the moose build for clang 11 and fixed up some others in this PR. We will still see some warnings related to libmesh contribs that will hang around for now, after my conversations with Roy in libmesh/libmesh#2812. So the next libMesh update should take care of some of what you're seeing, and this PR should take care of much of the rest.

**Comment by cticenhour** [comment]
That is, if the warnings we're speaking of are the same ones.

**Comment by cticenhour** [comment]
@lindsayad When would you like to do a libmesh update? I'd like to get testing started on this, and I know that it'll fail currently with some of our recipes that use `-Werror`.

**Comment by milljm** [comment]
We can't perform one yet. Not until our Firewall rules get fixed (See how master is unable to update any Anaconda packages: https://civet.inl.gov/branch/4/)

**Comment by cticenhour** [comment]
Right, we talked about that yesterday. Not looking for a merge due to that issue, but I was under the impression that modifications to conda recipes got PR tested. Was I wrong?

**Comment by cticenhour** [comment]
Or are the stacks not built for the first time until next?

**Comment by milljm** [comment]
You are correct, you can create PR's all you like. I am just stating: don't merge that into next until we fix firewall issues.

**Comment by lindsayad** [comment]
There are a few things I would like to see resolved before a libmesh update: libmesh/libmesh#2805 (@fdkong) and libmesh/libmesh#2749 (@loganharbour)

**Comment by loganharbour** [comment]
Thanks @lindsayad. libMesh/libmesh#2749 should be good this afternoon.

**Comment by cticenhour** [comment]
Sounds good - thanks for the updates everybody.

**Comment by fdkong** [comment]
> There are a few things I would like to see resolved before a libmesh update: [libMesh/libmesh#2805](https://github.com/libMesh/libmesh/pull/2805) (@fdkong) and [libMesh/libmesh#2749](https://github.com/libMesh/libmesh/pull/2749) (@loganharbour)

My libmesh PR is closed because I do not like the shared-pointer idea, and you guys free to go ahead for libmesh update

File: framework/contrib/gtest/gtest/gtest.h
Comment (cticenhour) [inline framework/contrib/gtest/gtest/gtest.h:21205]: @lindsayad I think this and Line 57 addresses your previous comment about this ignore impacting other unit test objects. The "push/pop" pragmas are the same used in the libmesh `ignore_warnings.h` and `restore_warnings.h` to isolate what is ignored.

**Comment by cticenhour** [comment]
@lindsayad I have rebased on the submodule updates you and Fande merged today, yet I am still seeing the `/usr/bin/bash` issue. Packages are being built on CentOS 8 and tested on Ubuntu 16.04. See https://civet.inl.gov/job/646099/

**Comment by lindsayad** [comment]
Ok yea I incorrectly identified the issue. Here’s what I think is going on: is this the first time we’ve built mpich on meowbuild @milljm ? I bet it is and mpich has /usr/bin stuff built into it now. I don’t have  any workarounds for that one. At the point it’s in the mpi compilers we’re pretty well done for

**Comment by milljm** [comment]
This is the first time we built the compilers on CentOS-7.7. Looks like we are going back to building the package on CentOS 7.6.

**Comment by cticenhour** [comment]
I haven't taken a look at the conda build and configuration scripts yet for this case, but are there edits we can make to get the mpich compiler package to look in the right place? Or would this be something deeper in this particular version of mpich?

**Comment by lindsayad** [comment]
This isn't something you can fix on the MPI side. It's all about PATH ordering on the configure/build system. If `/usr/bin` is ahead of `/bin` in PATH and `/usr/bin` has all the core utilities, then configure is going to pick up the `/usr/bin` utilities. And whereas `/bin` core utilities are guaranteed to exist on all systems (because a machine has to be able to work regardless of users), `/usr/bin` core utilities are not guaranteed to work on all systems

**Comment by cticenhour** [comment]
Right, I am aware of the distinction now between systems for this issue and that it is a configure/build issue - I was just wondering if there was some way we can use the configure or build scripts to standardize a bit for us without causing more headaches down the line. Like defining `PATH` more consistently in our build script to override system differences. I could see that one being dicey though.

**Comment by milljm** [comment]
I know this is confusing, but I think this PR fails to update moose-libmesh properly. Not that _this_ is the problem with building libmesh in the rest of the targets (I am looking into that). But this is what I end up with when I search for moose-libmesh after adding this PR's channel:

```
Loading channels: done
# Name                       Version           Build  Channel             
moose-libmesh             2020.06.02         build_0  idaholab            
moose-libmesh             2020.06.03         build_0  idaholab            
moose-libmesh             2020.06.16         build_0  idaholab            
moose-libmesh             2020.06.16         build_1  idaholab            
moose-libmesh             2020.07.23         build_0  idaholab            
moose-libmesh             2020.08.06         build_0  idaholab            
moose-libmesh             2020.10.05         build_0  idaholab            
moose-libmesh             2020.11.04         build_0  idaholab            
moose-libmesh             2020.11.04         build_1  idaholab            
moose-libmesh             2020.12.04         build_1  idaholab            
moose-libmesh             2020.12.10         build_0  moose/pr-16295      
moose-libmesh             2020.12.10         build_0  idaholab 
```

As you can see, there will be a conflict. Well, not really a conflict, but this PR's version of libMesh will overwrite the current one that is out there _not_ built with LLVM 11.

I don't have a good way of always keeping the last-PR-to-make-into-next, knowing when to bump the version number. It was either Alex's or Fande's libMesh update that is now 2020.12.10 build_0.

**Comment by cticenhour** [comment]
@milljm I imagine there's not a really good way of keeping that consistent beyond being vigilant....I'll go ahead and increment the build number for the YAML file so we can avoid the overwrite in this case.

**Comment by fdkong** [comment]
> Right, I am aware of the distinction now between systems for this issue and that it is a configure/build issue - I was just wondering if there was some way we can use the configure or build scripts to standardize a bit for us without causing more headaches down the line. Like defining `PATH` more consistently in our build script to override system differences. I could see that one being dicey though.

Nobody ever try to set an environment  like "SHELL=bin/bash"? Before we build compilers

**Comment by milljm** [comment]
I thought about adjusting path when ./build.sh is run. But I am not sure what that might do to Conda before it hands control over to `build.sh`.

I suppose all we can do is try it?

```bash
export PATH=/usr/bin:$PATH
```
at the top of each `build.sh`.

**Comment by lindsayad** [comment]
I'd be interested to give that a whirl

**Comment by cticenhour** [comment]
Seems to build without a hitch, but doesn't make a difference at testing time....https://civet.inl.gov/job/646684/

**Comment by cticenhour** [comment]
Can conda packages edit `PATH` when you activate the environment? Perhaps that's the way.

**Comment by cticenhour** [comment]
Interestingly, this works on another Ubuntu 16 instance https://civet.inl.gov/job/646693/

**Comment by milljm** [comment]
I am not quite understanding what is slowing down the non-meowbuild machines when dealing with Conda. +1hr seems common now. Even on our Trashcan machines, it seems to be taking longer.

**Comment by cticenhour** [comment]
Just judging from the intermittent failures I saw late yesterday, it seems like the cleaning step we have in place just isn't enough sometimes. One invalidation caught an "unrecoverable conda issue" and did a hard reset successfully, but then the other performed a standard cleaning and proceeded.

**Comment by milljm** [comment]
You wouldn't have a link to both of these events would you?

**Comment by cticenhour** [comment]
Here you go: 

- This one failed, Fande invalidated late last night, and it passed after a reinitialize and cleaning. Unfortunately I do not have a record of the previous results before the invalidation: https://civet.inl.gov/job/646684/
- This one failed, I invalidated it yesterday afternoon, and it passed after experiencing a conda issue and performing a reset during the Initialize step: https://civet.inl.gov/job/646692/
- This passed on the first go, but did require the same reset during Initialize:  https://civet.inl.gov/job/646693/

**Comment by milljm** [comment]
I'll add a print statement to the diff command that detects a 'tainted' environment. See if we can't spot why this happens all the time.

File: framework/contrib/gtest/gtest/gtest.h
Comment (lindsayad) [inline framework/contrib/gtest/gtest/gtest.h:62]: This is exactly what I was hoping to see. Thanks!

**Review by lindsayad — APPROVED** [review]
LGTM!

File: framework/contrib/gtest/gtest/gtest.h
Comment (cticenhour) [reply framework/contrib/gtest/gtest/gtest.h:62]: Thanks for prompting the change! I didn't know about these pragmas until I dug into those libmesh headers a bit more! :D

**Comment by cticenhour** [comment]
FYI - holding merge for now until I have a go from Dev-Ops folks.

**Review by milljm — APPROVED** [review]
👍 
LLVM 11! Nice!

**Pull request merged**
